File: FarNet/FarNetMan/Registry.cpp
/*
FarNet plugin for Far Manager
Copyright (c) 2005 FarNet Team
*/

#include "StdAfx.h"
#include "Registry.h"
using namespace Microsoft::Win32;

namespace FarNet
{;
String^ FarRegistryKey::RegistryPath::get()
{
	if (!_RegistryPath_)
	{
		String^ path = gcnew String(Info.RootKey);
		_RegistryPath_ = path->Substring(0, path->LastIndexOf('\\'));
	}

	return _RegistryPath_;
}

FarRegistryKey::FarRegistryKey(RegistryKey^ key)
: Key(key)
{}

FarRegistryKey::~FarRegistryKey()
{
	Key->Close();
}

FarRegistryKey::!FarRegistryKey()
{
	Key->Close();
}

String^ FarRegistryKey::Name::get()
{
	return Key->Name;
}

int FarRegistryKey::SubKeyCount::get()
{
	return Key->SubKeyCount;
}

int FarRegistryKey::ValueCount::get()
{
	return Key->ValueCount;
}

array<String^>^ FarRegistryKey::GetSubKeyNames()
{
	return Key->GetSubKeyNames();
}

array<String^>^ FarRegistryKey::GetValueNames()
{
	return Key->GetValueNames();
}

Object^ FarRegistryKey::GetValue(String^ name, Object^ defaultValue)
{
	return Key->GetValue(name, defaultValue);
}

void FarRegistryKey::SetValue(String^ name, Object^ value)
{
	if (!value)
		Key->DeleteValue(name, false);
	else if (value->GetType() == Int64::typeid)
		Key->SetValue(name, value, RegistryValueKind::QWord);
	else
		Key->SetValue(name, value);
}

Object^ FarRegistryKey::GetFarValue(String^ path, String^ valueName, Object^ defaultValue)
{
	RegistryKey^ key = nullptr;
	try
	{
		key = Registry::CurrentUser->OpenSubKey(RegistryPath + "\\" + path);
		return key ? key->GetValue(valueName, defaultValue) : defaultValue;
	}
	finally
	{
		delete key;
	}
}

void FarRegistryKey::SetFarValue(String^ path, String^ valueName, Object^ newValue)
{
	RegistryKey^ key = nullptr;
	try
	{
		key = Registry::CurrentUser->CreateSubKey(RegistryPath + "\\" + path);
		key->SetValue(valueName, newValue);
	}
	finally
	{
		if (key)
			key->Close();
	}
}

void FarRegistryKey::DeleteSubKey(String^ subkey)
{
	Key->DeleteSubKey(subkey, false);
}

IRegistryKey^ FarRegistryKey::OpenRegistryKey(String^ name, bool writable)
{
	// try to get existing
	String^ path = ES(name) ? RegistryPath : RegistryPath + "\\" + name;
	RegistryKey^ r = Registry::CurrentUser->OpenSubKey(path, writable);
	
	// create, throw
	if (!r && writable)
	{
		r = Registry::CurrentUser->CreateSubKey(path);
		if (!r)
			throw gcnew ModuleException("Cannot open the registry key.");
	}

	// ok
	return r ? gcnew FarRegistryKey(r) : nullptr;
}

}

File: tags/4.3.15/trunk/FarNet/FarNet/Registry.cs

/*
FarNet plugin for Far Manager
Copyright (c) 2005 FarNet Team
*/

using System;

namespace FarNet
{
	/// <summary>
	/// Virtual registry key.
	/// </summary>
	/// <remarks>
	/// Any opened key has to be disposed by <c>Dispose()</c>.
	/// <para>
	/// This interface is the simplified version of <c>Microsoft.Win32.RegistryKey</c>.
	/// In the Far Manager host it works with the real Windows registry.
	/// But modules should not rely on this anyway.
	/// </para>
	/// </remarks>
	public interface IRegistryKey : IDisposable
	{
		/// <summary>
		/// Retrieves the name of the key.
		/// </summary>
		string Name { get; }
		/// <summary>
		/// Retrieves the count of subkeys of the current key.
		/// </summary>
		[System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Naming", "CA1702:CompoundWordsShouldBeCasedCorrectly")]
		int SubKeyCount { get; }
		/// <summary>
		/// Retrieves the count of values in the key.
		/// </summary>
		int ValueCount { get; }
		/// <summary>
		/// Retrieves the value associated with the specified name. If the name is not found, returns the default value that you provide.
		/// </summary>
		object GetValue(string name, object defaultValue);
		/// <summary>
		/// Sets or deletes the specified name\value pair.
		/// </summary>
		/// <remarks>
		/// Setting the null value deletes the name\value pair.
		/// </remarks>
		void SetValue(string name, object value);
		/// <summary>
		/// Retrieves an array of strings that contains all the subkey names.
		/// </summary>
		[System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Naming", "CA1702:CompoundWordsShouldBeCasedCorrectly")]
		string[] GetSubKeyNames();
		/// <summary>
		/// Retrieves an array of strings that contains all the value names associated with this key.
		/// </summary>
		string[] GetValueNames();
		/// <summary>
		/// Deletes the specified subkey.
		/// </summary>
		/// <param name="subkey">The name of the subkey to delete.</param>
		/// <remarks>
		/// *) The subkey with its own subkeys cannot be deleted.
		/// *) It is OK if the specified subkey does not exit.
		/// </remarks>
		[System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Naming", "CA1702:CompoundWordsShouldBeCasedCorrectly")]
		void DeleteSubKey(string subkey);
	}
}
